src/oidcop/oauth2/token.py

```diff
@@ -358,6 +362,284 @@ def post_parse_request(
+
+        return resp
+
+    def _enforce_policy(self, request, token, config):
+        _context = self.endpoint.server_get("endpoint_context")
+
+        subject_token_types_supported = (
+            self.total_subject_token_types_supported.keys()
+            & config.get("subject_token_types_supported", "urn:ietf:params:oauth:token-type:access_token")
+        )
+        subject_token_types_supported = {
+            k:self.total_subject_token_types_supported[k] for k in subject_token_types_supported
+        }
+
+        if (
+            request["subject_token_type"] in subject_token_types_supported
+            and (
+                subject_token_types_supported[request["subject_token_type"]] != token.token_class
+            )
+        ):
+            return self.error_cls(
+                error="invalid_request", error_description="Wrong token type"
+            )
+
+        if request["subject_token_type"] not in subject_token_types_supported:
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported subject token type",
+            )
+
+        if (
+            "requested_token_type" in request
+            and request["requested_token_type"] not in config["requested_token_types_supported"]
+        ):
+            return TokenErrorResponse(
+                error="invalid_request",
+                error_description="Unsupported requested token type",
+            )
+
+        request_info = dict(scope=request.get("scope", []))
+        try:
+            check_unknown_scopes_policy(request_info, request["client_id"], _context)
+        except UnAuthorizedClientScope:
+            logger.error("Unauthorized scope requested.")
+            return self.error_cls(
+                error="invalid_grant",
+                error_description="Unauthorized scope requested",
+            )
+
+        try:
+            subject_token_type = request.get("subject_token_type", "")
+            if subject_token_type not in config["policy"]:
+                subject_token_type = ""
+            callable = config["policy"][subject_token_type]["callable"]
+            kwargs = config["policy"][subject_token_type]["kwargs"]
+
+            if isinstance(callable, str):
+                fn = importer(callable)
+            else:
+                fn = callable
+            return fn(request, context=_context, subject_token=token, **kwargs)
+
+        except Exception:
+            return self.error_cls(
+                error="server_error",
+                error_description="Internal server error"
+            )
+
+    def token_exchange_response(self, token):
+        response_args = {}
+        response_args["access_token"] = token.value
+        response_args["scope"] = token.scope
+        response_args["issued_token_type"] = token.token_class
+        response_args["expires_in"] = token.usage_rules.get("expires_in", 0)
+        response_args["token_type"] = "bearer"
```

Comment:
From https://datatracker.ietf.org/doc/html/rfc8693#section-2.2.1:

> token_type
   ...
   If the issued token is not an access token or 
   usable as an access token, then the "token_type" value "N_A" is
   used to indicate that an OAuth 2.0 "token_type" identifier is not
   applicable in that context.